**Windows** [Build 1111](http://brontes.lab.eng.brq.redhat.com/viewLog.html?buildId=37652&buildTypeId=WF_PullRequest_Windows) outcome was **FAILURE** using a merge of bdfa42581aaa1e33922bdae7ca2dc454b6d26eda
Summary: Tests failed: 2 (2 new), passed: 2927, ignored: 314 Build time: 0:48:46
### Build problems:
Failed tests detected

### Failed tests
```
org.jboss.as.test.integration.ejb.stateful.passivation.PassivationTestCase.testPassivationMaxSize: javax.ejb.EJBException: java.lang.IllegalStateException: WFLYCLEJBINF0008: Stateful session bean {[-45, -71, 37, -18, 104, 24, 75, -51, -92, 51, 125, 2, 96, -86, -106, -26]} refers to an invalid bean group 370cdc0b-cca6-415f-b609-f83e71f9a3d0
	at org.wildfly.clustering.ejb.infinispan.bean.InfinispanBeanFactory.createBean(InfinispanBeanFactory.java:76)
	at org.wildfly.clustering.ejb.infinispan.InfinispanBeanManager.findBean(InfinispanBeanManager.java:266)

org.jboss.as.test.manualmode.ws.ReloadRequiringChangesTestCase.testWSDLHostUndefineRequiresReloadAndDoesNotAffectRuntime: java.lang.AssertionError: null
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)

```

--- a/webservices/server-integration/src/main/java/org/jboss/as/webservices/deployers/WSIntegrationProcessorJAXWS_HANDLER.java
+++ b/webservices/server-integration/src/main/java/org/jboss/as/webservices/deployers/WSIntegrationProcessorJAXWS_HANDLER.java
@@ -93,7 +93,9 @@ public final class WSIntegrationProcessorJAXWS_HANDLER extends AbstractIntegrati
                 final String endpointClassName = classInfo.name().toString();
                 final ConfigResolver configResolver = new ConfigResolver(classInfo, jbossWebservicesMD, jwmd, root, war);
                 final EndpointConfig config = configResolver.resolveEndpointConfig();
-                registerConfigMapping(endpointClassName, config, unit);
+                if (config != null) {
+                    registerConfigMapping(endpointClassName, config, unit);
+                }
                 final Set<String> handlers = getHandlers(endpointClassName, config, configResolver, mapping);
                 if (!handlers.isEmpty()) {
                     if (isEjb3(classInfo)) {

--- a/webservices/server-integration/src/main/java/org/jboss/as/webservices/injection/WSHandlerChainAnnotationProcessor.java
+++ b/webservices/server-integration/src/main/java/org/jboss/as/webservices/injection/WSHandlerChainAnnotationProcessor.java
@@ -1,6 +1,6 @@
 /*
 * JBoss, Home of Professional Open Source.
-* Copyright 2011, Red Hat, Inc., and individual contributors
+* Copyright 2015, Red Hat, Inc., and individual contributors
 * as indicated by the @author tags. See the copyright.txt file in the
 * distribution for a full listing of individual contributors.
 *
@@ -46,11 +46,13 @@ import org.jboss.as.server.deployment.DeploymentUnit;
 import org.jboss.as.server.deployment.DeploymentUnitProcessingException;
 import org.jboss.as.server.deployment.DeploymentUnitProcessor;
 import org.jboss.as.server.deployment.module.ResourceRoot;
-import org.jboss.as.webservices.util.ASHelper;
 import org.jboss.as.webservices.logging.WSLogger;
+import org.jboss.as.webservices.util.ASHelper;
 import org.jboss.jandex.AnnotationInstance;
 import org.jboss.jandex.AnnotationTarget;
+import org.jboss.jandex.AnnotationValue;
 import org.jboss.jandex.ClassInfo;
+import org.jboss.jandex.DotName;
 import org.jboss.jandex.Index;
 import org.jboss.vfs.VirtualFile;
 import org.jboss.wsf.spi.metadata.j2ee.serviceref.UnifiedHandlerChainMetaData;
@@ -59,7 +61,10 @@ import org.jboss.wsf.spi.metadata.j2ee.serviceref.UnifiedHandlerChainsMetaDataPa
 import org.jboss.wsf.spi.metadata.j2ee.serviceref.UnifiedHandlerMetaData;
 
 /**
+ * Scans @HandlerChain annotations in the deployment
+ *
  * @author <a href="mailto:ropalka@redhat.com">Richard Opalka</a>
+ * @author <a href="mailto:alessio.soldano@jboss.com">Alessio Soldano</a>
  */
 public final class WSHandlerChainAnnotationProcessor implements DeploymentUnitProcessor {
 
@@ -94,16 +99,23 @@ public final class WSHandlerChainAnnotationProcessor implements DeploymentUnitPr
     }
 
     private static void processHandlerChainAnnotations(final ResourceRoot resourceRoot, final Index index, final WSEndpointHandlersMapping mapping) throws DeploymentUnitProcessingException {
-        final List<AnnotationInstance> handlerChainAnnotations = index.getAnnotations(HANDLER_CHAIN_ANNOTATION);
-
-        for (final AnnotationInstance handlerChainAnnotation : handlerChainAnnotations) {
-            final AnnotationTarget annotationTarget = handlerChainAnnotation.target();
-
+        final List<AnnotationInstance> webServiceAnnotations = index.getAnnotations(WEB_SERVICE_ANNOTATION);
+        final List<AnnotationInstance> webServiceProviderAnnotations = index.getAnnotations(WEB_SERVICE_PROVIDER_ANNOTATION);
+        for (final AnnotationInstance annotationInstance : webServiceAnnotations) {
+            final AnnotationTarget annotationTarget = annotationInstance.target();
             if (annotationTarget instanceof ClassInfo) {
                 final ClassInfo classInfo = (ClassInfo) annotationTarget;
                 if (isJaxwsEndpoint(classInfo, index)) {
-                    final String endpointClass = classInfo.name().toString();
-                    processHandlerChainAnnotation(resourceRoot, handlerChainAnnotation, endpointClass, mapping);
+                    AnnotationInstance handlerChainAnnotationInstance = getHandlerChainAnnotationInstance(classInfo);
+                    //JSR-181, Section 4.6.1: "The @HandlerChain annotation MAY be present on the endpoint interface and service
+                    //implementation bean. The service implementation bean’s @HandlerChain is used if @HandlerChain is present on both."
+                    if (handlerChainAnnotationInstance == null) {
+                        handlerChainAnnotationInstance = getEndpointInterfaceHandlerChainAnnotationInstance(classInfo, index);
+                    }
+                    if (handlerChainAnnotationInstance != null) {
+                        final String endpointClass = classInfo.name().toString();
+                        processHandlerChainAnnotation(resourceRoot, handlerChainAnnotationInstance, endpointClass, mapping);
+                    }
                 }
             } else {
                 // We ignore fields & methods annotated with @HandlerChain.
@@ -111,6 +123,40 @@ public final class WSHandlerChainAnnotationProcessor implements DeploymentUnitPr
                 // which are always referencing JAXWS client proxies only.
             }
         }
+        for (final AnnotationInstance annotationInstance : webServiceProviderAnnotations) {
+            final AnnotationTarget annotationTarget = annotationInstance.target();
+            if (annotationTarget instanceof ClassInfo) {
+                final ClassInfo classInfo = (ClassInfo) annotationTarget;
+                final AnnotationInstance handlerChainAnnotationInstance = getHandlerChainAnnotationInstance(classInfo);
+                if (handlerChainAnnotationInstance != null && isJaxwsEndpoint(classInfo, index)) {
+                    final String endpointClass = classInfo.name().toString();
+                    processHandlerChainAnnotation(resourceRoot, handlerChainAnnotationInstance, endpointClass, mapping);
+                }
+            } else {
+                // We ignore fields & methods annotated with @HandlerChain.
+                // These are used always in combination with @WebServiceRef
+                // which are always referencing JAXWS client proxies only.
+            }
+        }
+    }
+
+    private static AnnotationInstance getHandlerChainAnnotationInstance(final ClassInfo classInfo) {
+        List<AnnotationInstance> list = classInfo.annotations().get(HANDLER_CHAIN_ANNOTATION);
+        return list != null && !list.isEmpty() ? list.iterator().next() : null;
+    }
+
+    private static AnnotationInstance getEndpointInterfaceHandlerChainAnnotationInstance(final ClassInfo classInfo, final Index index) {
+        AnnotationValue av = classInfo.annotations().get(WEB_SERVICE_ANNOTATION).iterator().next().value("endpointInterface");
+        if (av != null) {
+            String intf = av.asString();
+            if (intf != null && !intf.isEmpty()) {
+                ClassInfo intfClassInfo = index.getClassByName(DotName.createSimple(intf));
+                if (intfClassInfo != null && ASHelper.isJaxwsEndpointInterface(intfClassInfo)) {
+                    return getHandlerChainAnnotationInstance(intfClassInfo);
+                }
+            }
+        }
+        return null;
     }
 
     private static void processHandlerChainAnnotation(final ResourceRoot resourceRoot, final AnnotationInstance handlerChainAnnotation, final String endpointClass, final WSEndpointHandlersMapping mapping) throws DeploymentUnitProcessingException {
@@ -177,5 +223,4 @@ public final class WSHandlerChainAnnotationProcessor implements DeploymentUnitPr
         final boolean isWebServiceProvider = clazz.annotations().containsKey(WEB_SERVICE_PROVIDER_ANNOTATION);
         return isWebService || isWebServiceProvider;
     }
-
 }

--- a/webservices/server-integration/src/main/java/org/jboss/as/webservices/util/ASHelper.java
+++ b/webservices/server-integration/src/main/java/org/jboss/as/webservices/util/ASHelper.java
@@ -198,6 +198,13 @@ public final class ASHelper {
         return false;
     }
 
+    public static boolean isJaxwsEndpointInterface(final ClassInfo clazz) {
+        final short flags = clazz.flags();
+        if (!Modifier.isInterface(flags)) return false;
+        if (!Modifier.isPublic(flags)) return false;
+        return clazz.annotations().containsKey(WEB_SERVICE_ANNOTATION);
+    }
+
     public static boolean hasClassesFromPackage(final Index index, final String pck) {
         for (ClassInfo ci : index.getKnownClasses()) {
             if (ci.name().toString().startsWith(pck)) {

--- a/webservices/server-integration/src/main/java/org/jboss/as/webservices/util/ServiceContainerEndpointRegistry.java
+++ b/webservices/server-integration/src/main/java/org/jboss/as/webservices/util/ServiceContainerEndpointRegistry.java
@@ -79,13 +79,11 @@ public class ServiceContainerEndpointRegistry implements EndpointRegistry {
         return false;
     }
 
-    @Override
     public void register(Endpoint endpoint) {
         // TODO:Remove this interface
 
     }
 
-    @Override
     public void unregister(Endpoint endpoint) {
         // TODO:Remove this interface
     }

